Triton data converter (CMSSW_11_2_0_pre9)

drankincms opened the pull request:
This is a first attempt at implementing a convertor factory for transforming inputs from one data type to another, useful for running on different servers with different precision without recompile the main producer.

Converters are located in `HeterogeneousCore/SonicTriton/plugins/converters/`, and currently two trivial converters are included (FloatStandardConverter.cc and Int64StandardConverter.cc) for `float` and `int64_t` in addition to an example converter for `float` to `ap_fixed<16,6>` (FloatApFixed16Converter.cc).

kpedro88 commented on HeterogeneousCore/SonicTriton/src/TritonData.cc:
```diff
@@ -116,14 +130,16 @@ void TritonInputData::toServer(std::shared_ptr<TritonInput<DT>> ptr) {
-    throw cms::Exception("TritonDataError") << name_ << " input(): inconsistent byte size " << sizeof(DT)
+  std::unique_ptr<TritonConverterBase<DT>> converter = createConverter<DT>();
+
+  if (byteSize_ != converter->getByteSize())
```
I'm not sure if this change is correct. The original purpose here was a runtime check that the type in the user's code (`DT`) corresponds to the model specification on the server (`byteSize_`). The converter can have a different byte size in the FPGA case, but is that what the server will return from the model specification?

drankincms replied:
I think this is necessary. I think the model specification on the server should use a type with the appropriate bitwidth (i.e. for the ap_fixed<16,6> FPGA setup the server returns UINT16 for the data type, and for a FP16 GPU model I think something similar would get done). But the user wants to just load the data as float32s in these cases.

kpedro88 commented on HeterogeneousCore/SonicTriton/src/TritonData.cc:
```diff
@@ -116,14 +130,16 @@ void TritonInputData::toServer(std::shared_ptr<TritonInput<DT>> ptr) {
 
-  if (byteSize_ != sizeof(DT))
-    throw cms::Exception("TritonDataError") << name_ << " input(): inconsistent byte size " << sizeof(DT)
+  std::unique_ptr<TritonConverterBase<DT>> converter = createConverter<DT>();
```
actually, this could also simplify the implementation of `createConverter()` somewhat. it wouldn't need to have a return type, because it could just store the converter in the `std::any` member variable directly.

drankincms replied:
so I thought a bit about this and I think we run into the issue of needing to know the data type at a point when the code is currently not aware of it. This is because `createConverter()` (or really the actual factory `get()` inside it) needs to know the type of factory. So we have to call `createConverter()` inside `toServer()` or `fromServer()`, and this seems like it defeats the purpose of using `std::any` here. Please correct me if Im wrong

kpedro88 replied:
The point is to avoid actually re-constructing the converter every time. Just make it once, then store it as a class member and reuse it. `std::any` allows storing as a class member.

drankincms replied:
Right, but I want to do this in the constructor presumably, which doesn't yet know the data type, so I can't call createConverter since it's templated. Or is the suggestion here to play some game to only create the converter the first time `toServer()` is called?

kpedro88 replied:
Yes: check if the `any` is null when cast to the correct type, and then create it, otherwise use the existing one.

drankincms replied:
Ah, ok, this makes sense to me now, thanks. In order to do this I'll need to make `fromServer` a nonconst function, since it needs to be able to create the converter. I want to make sure you agree with this change before I go ahead

kpedro88 replied:
Ah, that does break the interface a bit. Let's try making the member variable mutable instead; that should be acceptable here.

drankincms replied:
ok, sounds good

kpedro88 commented on HeterogeneousCore/SonicTriton/plugins/BuildFile.xml:
```diff
@@ -0,0 +1,11 @@
+  <use   name="FWCore/Utilities"/>
+  <use   name="FWCore/ParameterSet"/>
+  <use   name="FWCore/PluginManager"/>
+  <use   name="DataFormats/Common"/>
```
this include should go in the package BuildFile (rather than the plugin BuildFile) since it's used in the .h file

kpedro88 commented on HeterogeneousCore/SonicTriton/plugins/BuildFile.xml:
```diff
@@ -0,0 +1,11 @@
+  <use   name="FWCore/ParameterSet"/>
+  <use   name="FWCore/PluginManager"/>
+  <use   name="DataFormats/Common"/>
+  <use   name="HeterogeneousCore/SonicTriton"/>
```
I think this is really the only CMSSW include needed in the plugin BuildFile (the ones from the package BuildFile don't need to be repeated)

kpedro88 commented on HeterogeneousCore/SonicTriton/plugins/BuildFile.xml:
```diff
@@ -0,0 +1,11 @@
+<library   name="HeterogeneousCoreSonicTritonPlugins_converters" file="converters/*.cc">
```
recommended style is now to put only one space between xml attributes (everywhere)

kpedro88 commented on HeterogeneousCore/SonicTriton/plugins/BuildFile.xml:
```diff
@@ -0,0 +1,11 @@
+  <use   name="HeterogeneousCore/SonicTriton"/>
+  <use   name="triton-inference-server"/>
+  <use   name="protobuf"/>
+  <use   name="hls"/>
```
and this should be the only external include needed in this BuildFile

kpedro88 commented on HeterogeneousCore/SonicTriton/interface/TritonConverterBase.h:
```diff
@@ -0,0 +1,37 @@
+  virtual const uint8_t* convertIn(const DT* in) = 0;
+  virtual const DT* convertOut(const uint8_t* in) = 0;
+
+  virtual const int64_t getByteSize() const { return byteSize_; }
```
also, it should be named `byteSize()` to follow CMS naming rules for getters

kpedro88 commented on HeterogeneousCore/SonicTriton/plugins/converters/FloatApFixed16Converter.cc:
```diff
@@ -0,0 +1,37 @@
+DEFINE_EDM_PLUGIN(TritonConverterFactory<float>, FloatApFixed16Converter<6>, "FloatApFixed16F6Converter");
+
+template <int I>
+const uint8_t* FloatApFixed16Converter<I>::convertIn(const float* in) {
```
these could be defined inline (more common for class templates)

kpedro88 commented on HeterogeneousCore/SonicTriton/interface/TritonConverterBase.h:
```diff
@@ -0,0 +1,37 @@
+  virtual ~TritonConverterBase() = default;
+  TritonConverterBase& operator=(const TritonConverterBase&) = delete;
+
+  virtual const uint8_t* convertIn(const DT* in) = 0;
```
can these functions be `const`?

kpedro88 commented on HeterogeneousCore/SonicTriton/plugins/converters/FloatStandardConverter.cc:
```diff
@@ -0,0 +1,15 @@
+
+DEFINE_EDM_PLUGIN(TritonConverterFactory<float>, FloatStandardConverter, "FloatStandardConverter");
+
+const uint8_t* FloatStandardConverter::convertIn(const float* in) { return reinterpret_cast<const uint8_t*>(in); }
```
these and the next ones for int64 can also probably be defined inline

drankincms commented on HeterogeneousCore/SonicTriton/interface/TritonData.h:
```diff
@@ -40,6 +42,13 @@ class TritonData {
 
+  void setConverterParams(const edm::ParameterSet& conf) {
+    converterConf_ = conf;
+    converterName_ = conf.getParameter<std::string>("converterName");
```
my original thought was that the bit specifications for ap_fixed could be given in the pset, but having now tried it there are some real difficulties with doing that. Ill remove it now and if somewhere down the line we need to add it back its pretty easy

drankincms replied:
Actually I spoke too soo, it looks like the factory `create()` function expects 2 args, so we are perhaps better off just keeping the basically empty pset around

kpedro88 replied:
The signature of that function is `create(const std::string& iName, Args... args)`. The first argument is used for lookup, while the second is variadic, and just has to match a constructor for the class to be built. For now, you could just pass the name rather than keeping the full PSet around.

drankincms replied:
Ok, thanks for that explanation, I didnt fully appreciate where `args` was being used. I have this working now

drankincms commented:
All comments are now addressed, and I have tested with `TritonImageProducer` and the singularity container. @kpedro88 Please let me know if you have comments on the final implementation of `std::any` to store the container or the `byteSize_` checks.

kpedro88 commented on HeterogeneousCore/SonicTriton/interface/TritonData.h:
```diff
@@ -47,7 +47,9 @@ class TritonData {
   template <typename DT>
-  std::unique_ptr<TritonConverterBase<DT>> createConverter() const { return TritonConverterFactory<DT>::get()->create(converterName_); }
+  void createConverter() const { 
+    if (!converter_.has_value()) converter_ = std::shared_ptr<TritonConverterBase<DT>>(TritonConverterFactory<DT>::get()->create(converterName_));
```
This checks if it's filled, but not if it's the right type. (I don't anticipate a given TritonData object being used for more than one primitive type, but we should try to be future-proof.) I propose the following (also to reduce type name repetition):
```cpp
  using ConverterType = std::shared_ptr<TritonConverterBase<DT>>;
  if (std::any_cast<ConverterType>(converter_)==nullptr) converter_ = ConverterType(TritonConverterFactory<DT>::get()->create(converterName_));
```

and actually, now I've realized that the member variable has to be `std::shared_ptr<std::any>`, because `any` has to contain a bare pointer in order for `any_cast` to return `nullptr` when the type is wrong. So it needs a little more revision than my above suggestion; let me know if this isn't clear.

drankincms replied:
In order to catch both the empty and wrong-type conversions, I think we want to keep `std::any` to store things. I tried with `std::shared_ptr<std::any>` and had some issues, and this seems to be what we need anyway

kpedro88 commented on HeterogeneousCore/SonicTriton/interface/TritonConverterBase.h:
```diff
@@ -0,0 +1,41 @@
+
+template <typename DT>
+class TritonConverterBase {
+//class needs to be templated since the convert functions require the data type, but need to also be virtual, and virtual member function templates are not allowed ion C++
```
typo ion -> in

drankincms replied:
Good catch :)

kpedro88 commented on HeterogeneousCore/SonicTriton/interface/TritonData.h:
```diff
@@ -40,6 +43,16 @@ class TritonData {
+  template <typename DT>
+  void createConverter() const { 
+    using ConverterType = std::shared_ptr<TritonConverterBase<DT>>;
+    //this contruction catches bad any_cast without throwing std exception
```
typo contruction -> construction

drankincms replied:
I need a spell checker clearly :)

kpedro88 commented on HeterogeneousCore/SonicTriton/plugins/converters/FloatApFixed16Converter.cc:
```diff
@@ -0,0 +1,32 @@
+    return reinterpret_cast<const uint8_t*>((this->makeVecIn(in)).data());
+  }
+  const float* convertOut(const uint8_t* in) const {
+    return (this->makeVecOut(reinterpret_cast<const ap_fixed<16, I>*>(in))).data();
```
I think this will encounter a lifetime issue. The object returned by `makeVecOut()` goes out of scope as soon as this function returns, so the pointer to its data could point to deallocated or reallocated memory.

In `TritonData`, I added a class member `std::any holder_` to extend the lifetime of the `TritonInput` object in `toServer()`. Something similar is probably needed here. It doesn't need to be `std::any` since the type is known, but it does need to be aware that the `convertIn/Out()` functions can be called multiple times per event (so all created vectors should be stored).

This implies it would be useful for `TritonConverterBase` to have a `clear()` function that can be called in `TritonData::reset()`. To handle the type erasure, it should be stored as an `std::function` member variable in `TritonData`, using `std::bind` when the converter is created.

drankincms replied:
So I have tried to set this up, but I think there may be an issue due to the fact that the converter base class is virtual, and the way that `std::function`/`std::bind` work. If I try to bid a non-virtual function I can get things to work, but since the clear function needs to be virtual (I think) I can't figure out a way to make it work. Perhaps I'm missing something about `std::bind`? I just wanted to comment here in case you have any suggestions, otherwise I will keep playing around.

kpedro88 replied:
As I understand it, storing a virtual member function in `std::function` should work if you give `std::bind` the pointer to the actual object (even if it's a base class pointer).

This test code worked for me:
```cpp
#include <iostream>
#include <functional>

class Base {
public:
    virtual void doIt() const { std::cout << "Base class" << std::endl; }
};

class Derived: public Base {
public:
    void doIt() const override { std::cout << "Derived class" << std::endl; }
};

int main(){
    Base* obj = new Derived();

    std::function<void()> f = std::bind(&Base::doIt, obj);
    f();
}
```

drankincms replied:
Thanks, I have tried this construction and in a few different attempts to implement it I always get a seg fault. Ill keep trying to work through it

kpedro88 replied:
I think it should only segfault if the object pointer gets lost/deallocated...

drankincms replied:
In the simplest case I use: `converter_clear = std::bind(&TritonConverterBase<DT>::clear, std::any_cast<std::shared_ptr<TritonConverterBase<DT>>>(converter_));` inside `createConverter()`. I can't really see a way that `converter_` should get lost, and indeed even trying to call `converter_clear()` right after I set it using `std::bind` gives the fault. Perhaps Im missing something obvious here?

kpedro88 replied:
the output of `std::any_cast<std::shared_ptr<TritonConverterBase<DT>>>(converter_))` has type `std::shared_ptr<TritonConverterBase<DT>`, which is not `TritonConverterBase<DT>*`. Try using `std::any_cast<std::shared_ptr<TritonConverterBase<DT>>>(converter_).get()` instead

drankincms replied:
I have tried both with and without the `.get()`, but both give the fault

kpedro88 replied:
Can you post the trace from the segfault?

drankincms replied:
`A fatal system signal has occurred: segmentation violation
The following is the call stack containing the origin of the signal.

Mon Nov 30 13:33:25 CST 2020
Thread 7 (Thread 0x7f586a3ff700 (LWP 22023)):
#0  0x00007f588bff6a35 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f586d70e6e2 in gpr_cv_wait () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#2  0x00007f586fe21a2f in timer_thread(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgrpc.so.9
#3  0x00007f586d710823 in grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::{lambda(void*)#1}::_FUN(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#4  0x00007f588bff2ea5 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f588bd1b96d in clone () from /lib64/libc.so.6
Thread 6 (Thread 0x7f586b1ff700 (LWP 22017)):
#0  0x00007f588bff6de2 in pthread_cond_timedwait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f586d70e74a in gpr_cv_wait () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#2  0x00007f586fe21a2f in timer_thread(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgrpc.so.9
#3  0x00007f586d710823 in grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::{lambda(void*)#1}::_FUN(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#4  0x00007f588bff2ea5 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f588bd1b96d in clone () from /lib64/libc.so.6
Thread 5 (Thread 0x7f586c3dc700 (LWP 22016)):
#0  0x00007f588bff6a35 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f586d70e6e2 in gpr_cv_wait () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#2  0x00007f586fe0e6fb in grpc_core::Executor::ThreadMain(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgrpc.so.9
#3  0x00007f586d710823 in grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::{lambda(void*)#1}::_FUN(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#4  0x00007f588bff2ea5 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f588bd1b96d in clone () from /lib64/libc.so.6
Thread 4 (Thread 0x7f586cbdd700 (LWP 22015)):
#0  0x00007f588bff6a35 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f586d70e6e2 in gpr_cv_wait () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#2  0x00007f586fe0e6fb in grpc_core::Executor::ThreadMain(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgrpc.so.9
#3  0x00007f586d710823 in grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::{lambda(void*)#1}::_FUN(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#4  0x00007f588bff2ea5 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f588bd1b96d in clone () from /lib64/libc.so.6
Thread 3 (Thread 0x7f586d3de700 (LWP 22014)):
#0  0x00007f588bff6a35 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f588c5d962c in __gthread_cond_wait (__mutex=<optimized out>, __cond=<optimized out>) at /data/cmsbld/jenkins/workspace/auto-builds/CMSSW_11_1_0_pre7-slc7_amd64_gcc820/build/CMSSW_11_1_0_pre7-build/BUILD/slc7_amd64_gcc820/external/gcc/8.2.0-bcolbf/gcc-8.2.0/obj/x86_64-unknown-linux-gnu/libstdc++-v3/include/x86_64-unknown-linux-gnu/bits/gthr-default.h:864
#2  std::condition_variable::wait (this=<optimized out>, __lock=...) at ../../../../../libstdc++-v3/src/c++11/condition_variable.cc:53
#3  0x00007f58701cb277 in SonicDispatcherPseudoAsync::waitForNext() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libHeterogeneousCoreSonicCore.so
#4  0x00007f588c5ded2f in std::execute_native_thread_routine (__p=0x7f587132d9a0) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#5  0x00007f588bff2ea5 in start_thread () from /lib64/libpthread.so.0
#6  0x00007f588bd1b96d in clone () from /lib64/libc.so.6
Thread 2 (Thread 0x7f5870e37700 (LWP 22003)):
#0  0x00007f588bffa1d9 in waitpid () from /lib64/libpthread.so.0
#1  0x00007f5886d6afd7 in edm::service::cmssw_stacktrace_fork() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#2  0x00007f5886d6ba9a in edm::service::InitRootHandlers::stacktraceHelperThread() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#3  0x00007f588c5ded2f in std::execute_native_thread_routine (__p=0x7f5881b6f1b0) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#4  0x00007f588bff2ea5 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f588bd1b96d in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x7f588a304540 (LWP 21976)):
#0  0x00007f588bd10c3d in poll () from /lib64/libc.so.6
#1  0x00007f5886d6b43f in full_read.constprop () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#2  0x00007f5886d6bb7c in edm::service::InitRootHandlers::stacktraceFromThread() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#3  0x00007f5886d6ca59 in sig_dostack_then_abort () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#4  <signal handler called>
#5  0x0000000000000000 in ?? ()
#6  0x00007f58701f2264 in void TritonData<nvidia::inferenceserver::client::InferInput>::toServer<float>(std::shared_ptr<std::vector<std::vector<float, std::allocator<float> >, std::allocator<std::vector<float, std::allocator<float> > > > >) () from /uscms_data/d3/drankin/Sonic/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libHeterogeneousCoreSonicTriton.so
#7  0x00007f5870215259 in TritonImageProducer::acquire(edm::Event const&, edm::EventSetup const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, TritonData<nvidia::inferenceserver::client::InferInput>, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, TritonData<nvidia::inferenceserver::client::InferInput> > > >&) () from /uscms_data/d3/drankin/Sonic/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/plugintestHeterogenousCoreSonicTriton.so
#8  0x00007f587020f55b in SonicAcquirer<TritonClient, edm::stream::EDProducer<edm::ExternalWork> >::acquire(edm::Event const&, edm::EventSetup const&, edm::WaitingTaskWithArenaHolder) () from /uscms_data/d3/drankin/Sonic/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/plugintestHeterogenousCoreSonicTriton.so
#9  0x00007f588e8bb319 in edm::stream::doAcquireIfNeeded(edm::stream::impl::ExternalWork*, edm::Event const&, edm::EventSetup const&, edm::WaitingTaskWithArenaHolder&) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#10 0x00007f588e8b994f in edm::stream::EDProducerAdaptorBase::doAcquire(edm::EventTransitionInfo const&, edm::ActivityRegistry*, edm::ModuleCallingContext const*, edm::WaitingTaskWithArenaHolder&) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#11 0x00007f588e886a62 in edm::Worker::runAcquire(edm::EventTransitionInfo const&, edm::ParentContext const&, edm::WaitingTaskWithArenaHolder&) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#12 0x00007f588e886bc1 in edm::Worker::runAcquireAfterAsyncPrefetch(std::__exception_ptr::exception_ptr const*, edm::EventTransitionInfo const&, edm::ParentContext const&, edm::WaitingTaskWithArenaHolder) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#13 0x00007f588e7fafb3 in edm::Worker::AcquireTask<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>, void>::execute() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#14 0x00007f588cfecbfd in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::process_bypass_loop (this=this@entry=0x7f58895b3200, context_guard=..., t=t@entry=0x7f58895a7340, isolation=isolation@entry=0) at ../../src/tbb/custom_scheduler.h:393
#15 0x00007f588cfecef5 in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all (this=0x7f58895b3200, parent=..., child=<optimized out>) at ../../include/tbb/task.h:1003
#16 0x00007f588e77b355 in edm::EventProcessor::processLumis(std::shared_ptr<void> const&) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#17 0x00007f588e78349e in edm::EventProcessor::runToCompletion() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#18 0x000000000040f59d in tbb::interface7::internal::delegated_function<main::{lambda()#1}::operator()() const::{lambda()#1} const, void>::operator()() const ()
#19 0x00007f588cfe7bc1 in tbb::interface7::internal::task_arena_base::internal_execute (this=0x7ffdbfdfaf50, d=...) at ../../src/tbb/arena.cpp:1105
#20 0x0000000000410488 in main::{lambda()#1}::operator()() const ()
#21 0x000000000040eff5 in main ()

Current Modules:

Module: TritonImageProducer:TritonProducer (crashed)

A fatal system signal has occurred: segmentation violation`

drankincms replied:
A clean build attempt seems to fix this issue, so the `clear()` is now implemented

kpedro88 commented on HeterogeneousCore/SonicTriton/src/TritonData.cc:
```diff
@@ -116,14 +117,16 @@ void TritonInputData::toServer(std::shared_ptr<TritonInput<DT>> ptr) {
 
-  if (byteSize_ != sizeof(DT))
-    throw cms::Exception("TritonDataError") << name_ << " input(): inconsistent byte size " << sizeof(DT)
+  createConverter<DT>();
```
to reduce the duplication of `std::any_cast<std::shared_ptr<TritonConverterBase<DT>>>(converter_)`, maybe `createConverter<DT>()` can do this cast and return the `std::shared_ptr<TritonConverterBase<DT>>`? Then this could be:
```cpp
auto converter = createConverter<DT>();
```

kpedro88 commented on HeterogeneousCore/SonicTriton/interface/TritonData.h:
```diff
@@ -40,6 +43,16 @@ class TritonData {
+    converterName_ = conf.getParameter<std::string>("converterName");
+  }
+  template <typename DT>
+  void createConverter() const { 
```
this function (and the above `setConverterParams()` should go in the `private` section of this class

kpedro88 commented on HeterogeneousCore/SonicTriton/interface/TritonData.h:
```diff
@@ -93,6 +112,9 @@ class TritonData {
   std::shared_ptr<Result> result_;
+  mutable std::any converter_;
+  std::string converterName_;
+  mutable std::function<void()> converter_clear;
```
member variables should end in underscore

kpedro88 commented on HeterogeneousCore/SonicTriton/plugins/converters/FloatApFixed16Converter.cc:
```diff
@@ -0,0 +1,44 @@
+    return temp_vec->data();
+  }
+
+  void clear() const {
```
this should be marked as override

kpedro88 replied:
(actually all the other inherited functions in the derived classes should as well)

kpedro88 commented on HeterogeneousCore/SonicTriton/src/TritonClient.cc:
```diff
@@ -336,10 +339,14 @@ inference::ModelStatistics TritonClient::getServerSideStatus() const {
 //for fillDescriptions
 void TritonClient::fillPSetDescription(edm::ParameterSetDescription& iDesc) {
+  edm::ParameterSetDescription descConverter;
+  fillBasePSetDescription(descConverter);
```
This line is unnecessary and should be removed

kpedro88 commented on HeterogeneousCore/SonicTriton/src/TritonClient.cc:
```diff
@@ -336,10 +339,14 @@ inference::ModelStatistics TritonClient::getServerSideStatus() const {
   fillBasePSetDescription(descClient);
   descClient.add<std::string>("modelName");
   descClient.add<std::string>("modelVersion", "");
+  descClient.add<edm::ParameterSetDescription>("converterDefinition", descConverter);
```
I keep thinking of this and forgetting to comment on it: a single converter definition per client isn't sufficient for the general case. Triton supports networks with multiple (named) inputs and outputs that can have different types. This will probably require two VPSets, `inputConverters` and `outputConverters` to be fully general, with each PSet in these VPSets having both the input/output name and the converter name.

drankincms replied:
Ok, good point. I finally have FaaST updated to v2 so that I can test properly, so I should be able to get this change included soon.

drankincms commented:
Ok, all the comments are now addressed, and I have changed things to allow for multiple separate converters.
This probably needs to be synced with parallel developments, since anything using SONIC will now need to have the converter definitions.

kpedro88 commented on HeterogeneousCore/SonicTriton/src/TritonClient.cc:
```diff
@@ -79,7 +79,7 @@ TritonClient::TritonClient(const edm::ParameterSet& params)
     throw cms::Exception("ModelErrors") << msg_str;
 
-  const edm::ParameterSet& converterDefs = params.getParameterSet("converterDefinition");
+  const std::vector<edm::ParameterSet>& inputConverterDefs = params.getParameter<std::vector<edm::ParameterSet>>("inputConverters");
```
a bit easier to use `getParameterSetVector()`

kpedro88 commented on HeterogeneousCore/SonicTriton/src/TritonClient.cc:
```diff
@@ -91,7 +91,14 @@ TritonClient::TritonClient(const edm::ParameterSet& params)
     auto& curr_input = curr_itr->second;
-    curr_input.setConverterParams(converterDefs);
+    bool foundConverter = false;
+    for (const auto converterDef : inputConverterDefs) {
```
This is essentially O(N^2) in the number of inputs N. Even though N is usually small, that is not ideal. It would be better to convert the initial VPSet into `unordered_map<string,string>` (inserting and searching both O(1) on average). (and do the same for the output section below)

kpedro88 commented on HeterogeneousCore/SonicTriton/test/tritonTest_cfg.py:
```diff
@@ -29,6 +29,40 @@
 }
 
+inConvs = {
+  "TritonImageProducer": cms.VPSet(
```
To simplify this for users, we might want to add some additional infrastructure:
1. To handle default cases, we could map the standard converters to the appropriate Triton data type names/enums (see the TritonData constructor). Then these wouldn't need to be specified in the python config explicitly.
2. We might want to investigate if there's a way to augment the model config and protocols so that, when a FaaST server is used, it could automatically specify (probably via some gRPC message) when an alternate converter needs to be used (and which one).

Item 2 assumes that there would not be a case where the client wants to use a different converter than specified by the model in the repository/on the server.

kpedro88 replied:
One idea to avoid a) overloading the meaning of some existing protobuf field (potentially dangerous) or b) extending the proto definitions (maintenance burden):
we could come up with a policy whereby the input and output names in the config.pbtxt are extended to indicate which non-standard converter (if any) is needed. Then these extended names could be handled by the client.
We could try to have the FaaST server perform this extension automatically (in order to keep the same config.pbtxt files for GPU and FPGA versions of a model). But since other model files might need to be different for FPGA usage, changing the config.pbtxt might be okay.

Even with this, it could still be useful to have the option introduced here to specify a converter manually (for testing).

drankincms replied:
1) Ok yeah, this is a good suggestion, Ill set this up.

2) Hmm, perhaps we could use something existing like the `platform` field in `ModelMetadataResponse`, and fill it with the server's desired converter. This seems simple enough for FaaST, but I'm not sure how easy this would be with a standard GPU Triton server.

kpedro88 replied:
There's still only one `platform` per model, so we would have to pack all the information for every input and output into that one string, which isn't ideal.

drankincms replied:
Oh yeah, I keep forgetting that other models have multiple inputs/outputs.

drankincms replied:
So this now uses defaults successfully when they are available, so 1) is addressed.

For 2), Ok, I missed your suggestion of the model name field. I think the version field might be even more accurate but it's a minor point. But can we enforce a policy whereby we use the field to encode the necessary converter? I'm wondering if you think this is actually feasible or if you were only thinking out loud a bit here.

kpedro88 replied:
To be a bit more explicit, the most automatic implementation would be:
1. config.pbtxt retains the normal input/output names (same for GPU and FPGA)
2. FaaST appends information to the input/output names when returning the `ModelMetadataResponse`, e.g. "input1_FPGAConverter:foo"
3. TritonClient knows to check for the presence of "_FPGAConverter:foo"

The question here would be how "foo" is provided for each model input/output. Maybe there could be a separate section of the config.pbtxt; we have to learn more about how Triton parses it and what (if anything) it ignores.

The alternative would be a bit more manual on the user side:
1. make a separate FPGA copy of the GPU config.pbtxt, and add "_FPGAConverter:foo" to the input/output names.
2. FaaST uses the FPGA copy of the config.pbtxt (maybe this could also be automated...)

kpedro88 commented on HeterogeneousCore/SonicTriton/interface/TritonData.h:
```diff
@@ -58,6 +61,17 @@ class TritonData {
 
+  std::string defaultConverter() const {
+    std::string base = "StandardConverter";
+    if (dname_ == "INT64") {
```
it would be slightly faster to use the enums (`dtype_`)
they're listed in https://github.com/triton-inference-server/server/blob/v2.3.0/src/core/model_config.cc#L40

kpedro88 commented on HeterogeneousCore/SonicTriton/interface/TritonData.h:
```diff
@@ -69,6 +83,22 @@ class TritonData {
   void setResult(std::shared_ptr<Result> result) { result_ = result; }
   IO* data() { return data_.get(); }
 
+  void setConverterParams(std::string convName) {
```
`const std::string&`

kpedro88 commented on HeterogeneousCore/SonicTriton/src/TritonClient.cc:
```diff
@@ -79,6 +79,12 @@ TritonClient::TritonClient(const edm::ParameterSet& params)
 
+  const std::vector<edm::ParameterSet>& inputConverterDefs = params.getParameterSetVector("inputConverters");
+  std::unordered_map<std::string,std::string> inConvMap;
+  for (const auto converterDef : inputConverterDefs) {
```
`const auto&`

kpedro88 commented on HeterogeneousCore/SonicTriton/src/TritonClient.cc:
```diff
@@ -101,6 +112,12 @@ TritonClient::TritonClient(const edm::ParameterSet& params)
 
+  const std::vector<edm::ParameterSet>& outputConverterDefs = params.getParameterSetVector("outputConverters");
+  std::unordered_map<std::string,std::string> outConvMap;
+  for (const auto converterDef : outputConverterDefs) {
```
`const auto&`

kpedro88 commented on HeterogeneousCore/SonicTriton/interface/TritonData.h:
```diff
@@ -58,6 +61,17 @@ class TritonData {
+    else if (dname_ == "FP32") {
+      return "Float"+base;
+    }
+    return "";
```
given the way this is currently used in the `TritonClient` constructor, it might be better for it to throw if it can't find a default converter (since the constructor already checked if a non-default converter was requested)
also, I'd recommend putting this in the private section (`TritonClient` is a friend class, so it can still call the function)

drankincms commented:
Ok, this got lost a bit during the holidays, just to refocus: @kpedro88 you had suggested modifying the name returned by FaaST to include the suggested data converter. Should that be added here before merging (also neglecting the necessary rebase presumably), or were you thinking that this feature would be included in a future PR?

kpedro88 commented:
@drankincms I think we should strive to include any foreseen and achievable automation in PRs to the official CMSSW repository. Doing everything by hand is fine for expert testing, but it's not really usable for other users in that mode, let alone for production.

It seems like the bulk of the changes will need to be made in the FaaST server code, with just some additional string parsing introduced in SonicTriton. I would definitely prefer to present a complete product for integration in CMSSW.

drankincms commented:
Ok, sounds good, Ill get back to this and add the feature now then 👍

kpedro88 commented on HeterogeneousCore/SonicTriton/src/TritonClient.cc:
```diff
@@ -79,17 +79,29 @@ TritonClient::TritonClient(const edm::ParameterSet& params)
   for (const auto& nicInput : nicInputs) {
-    const auto& iname = nicInput.name();
+    const std::string iname_full = nicInput.name();
+    const auto& iname = iname_full.find("_DataConverter:") != std::string::npos ? iname_full.substr(0,iname_full.find("_DataConverter:")) : iname_full;
```
here's how I would reduce duplication:

1. define a static function of TritonClient:
    ```cpp
    static std::pair<std::string, std::string> splitNameConverter(const std::string& fullname) {
      static const std::string indicator("_DataConverter:");
      size_t dcpos = fullname.find(indicator);
      if (dcpos != std::string::npos)
        return {fullname.substr(0,dcpos), fullname.substr(dcpos+indicator.size())};
      else
        return {fullname, ""};
    }
    ```
2. use the static function here:
    ```cpp
    const auto& [iname, iconverter] = TritonClient::splitNameConverter(iname_full);
    ```
3. pass `iconverter` to `curr_input.defaultConverter()` below (& do the same for the outputs further down)
4. `defaultConverter()` just has to check if the input parameter is non-empty, otherwise it tries to get a standard converter (as currently implemented)
5. `inConvMap` can be removed entirely (maybe not right away if you still want to test things)

drankincms replied:
Ok, yes, this is cleaner, Ill make this change now, thanks. Regarding 5), I think we still want to let the user define their own converter. For example, in the case of FACILE the server would be configured to request FloatToApFixed16F6 as the converter, but if a dedicated FACILE for FPGA producer is written, it can use the same server and would need to use a different converter.

kpedro88 replied:
I can imagine there might be cases where the user wants to provide an alternate converter, but I'm confused about your example. Why would a dedicated producer need to use a different converter for the same server?

drankincms replied:
Im envisioning a dedicated FPGA producer that is aware of the `ap_fixed<16,6>` type that is needed by the server, so it places each input in the data vector already converted, so the converter is basically just a default `UINT16` converter.

kpedro88 replied:
Ah okay, I understand now.

kpedro88 commented on HeterogeneousCore/SonicTriton/interface/TritonData.h:
```diff
@@ -69,6 +72,39 @@ class TritonData {
   IO* data() { return data_.get(); }
 
+  std::string defaultConverter(const std::string name) const {
+    if (name != "") {
```
more efficient: `if (!name.empty())`

drankincms replied:
Didn't know that, thanks :)

kpedro88 commented:
@drankincms let me know when FaaST is updated accordingly (we should test this before we put it into CMSSW)

drankincms commented:
Ok, just committed the changes for FaaST to work with v2 here: https://github.com/drankincms/FaaST/tree/v2
I can submit a PR to the fastml main repo if you'd like.

I have been testing this client code with this server, not sure if you're envisioning something more official before we put it in CMSSW.

kpedro88 commented:
yes, let's discuss the FaaST changes more in a PR to that repo.

it would be good if we had some recipe for other people (i.e. CMSSW PR reviewers) to test the code. I'm not sure if this is easy, though...

Eventually, we might even want to distribute Docker and Singularity containers for FaaST, as we do with Triton.

drankincms commented:
Sounds good. We can discuss ways we might be able to test, although it might be quite tricky, I don't know if there are many FPGA-connected resources available ATM.

Regarding the PR, do we want to keep separate branches there for v1 and v2 FaaST? I can make a new branch if this is preferable.

kpedro88 commented:
I don't see a reason to keep separate FaaST branches right now. The first version is tagged. If, for some reason, we need to go back and make a 1.x release, we can spin off a branch at that time.

drankincms commented:
Ok, done. https://github.com/fastmachinelearning/FaaST/pull/2
